File: intreview_training/design-patterns/creational/singleton/implementations/js/singleton.js
/**
 * Singleton Pattern - JavaScript Implementations
 * 
 * Multiple approaches to implement Singleton pattern in JavaScript
 */

// ============================================================================
// 1. EAGER INITIALIZATION
// ============================================================================
class EagerSingleton {
    // Instance created immediately when class is defined
    static instance = new EagerSingleton();

    // Private constructor (simulated with Symbol)
    constructor() {
        if (EagerSingleton.instance) {
            throw new Error('Use getInstance() instead of new');
        }
        this.data = 'Eager Singleton Data';
    }

    static getInstance() {
        return EagerSingleton.instance;
    }

    getData() {
        return this.data;
    }
}

// ============================================================================
// 2. LAZY INITIALIZATION
// ============================================================================
class LazySingleton {
    static instance = null;

    constructor() {
        if (LazySingleton.instance) {
            throw new Error('Use getInstance() instead of new');
        }
        this.data = 'Lazy Singleton Data';
    }

    static getInstance() {
        if (!LazySingleton.instance) {
            LazySingleton.instance = new LazySingleton();
        }
        return LazySingleton.instance;
    }

    getData() {
        return this.data;
    }
}

// ============================================================================
// 3. MODULE PATTERN (JavaScript-specific, simplest approach)
// ============================================================================
const ModuleSingleton = (function() {
    let instance;

    function createInstance() {
        return {
            data: 'Module Singleton Data',
            getData() {
                return this.data;
            }
        };
    }

    return {
        getInstance() {
            if (!instance) {
                instance = createInstance();
            }
            return instance;
        }
    };
})();

// ============================================================================
// 4. ES6 MODULE (Node.js/Modern JavaScript - Best Practice)
// ============================================================================
// In a separate file: logger.js
// let instance = null;
// 
// class Logger {
//     constructor() {
//         if (instance) {
//             return instance;
//         }
//         instance = this;
//         this.logs = [];
//     }
// 
//     log(message) {
//         this.logs.push(message);
//         console.log(message);
//     }
// }
// 
// export default new Logger(); // Export single instance

// ============================================================================
// 5. WITH PRIVATE FIELDS (ES2022)
// ============================================================================
class ModernSingleton {
    static #instance = null;

    #data = 'Modern Singleton Data';

    constructor() {
        if (ModernSingleton.#instance) {
            return ModernSingleton.#instance;
        }
        ModernSingleton.#instance = this;
    }

    static getInstance() {
        if (!ModernSingleton.#instance) {
            ModernSingleton.#instance = new ModernSingleton();
        }
        return ModernSingleton.#instance;
    }

    getData() {
        return this.#data;
    }
}

// ============================================================================
// 6. DATABASE CONNECTION EXAMPLE
// ============================================================================
class DatabaseConnection {
    static instance = null;
    static connectionCount = 0;

    constructor() {
        if (DatabaseConnection.instance) {
            return DatabaseConnection.instance;
        }
        
        // Simulate expensive connection setup
        this.connected = false;
        this.connectionString = '';
        DatabaseConnection.instance = this;
    }

    static getInstance() {
        if (!DatabaseConnection.instance) {
            DatabaseConnection.instance = new DatabaseConnection();
        }
        return DatabaseConnection.instance;
    }

    connect(connectionString) {
        if (this.connected) {
            console.log('Already connected');
            return this;
        }
        
        this.connectionString = connectionString;
        this.connected = true;
        DatabaseConnection.connectionCount++;
        console.log(`Connected to: ${connectionString}`);
        return this;
    }

    disconnect() {
        this.connected = false;
        console.log('Disconnected');
    }

    query(sql) {
        if (!this.connected) {
            throw new Error('Not connected to database');
        }
        console.log(`Executing: ${sql}`);
        return { rows: [] };
    }
}

// ============================================================================
// 7. LOGGER EXAMPLE
// ============================================================================
class Logger {
    static instance = null;
    logs = [];

    constructor() {
        if (Logger.instance) {
            return Logger.instance;
        }
        Logger.instance = this;
    }

    static getInstance() {
        if (!Logger.instance) {
            Logger.instance = new Logger();
        }
        return Logger.instance;
    }

    log(level, message) {
        const logEntry = {
            timestamp: new Date().toISOString(),
            level,
            message
        };
        this.logs.push(logEntry);
        console.log(`[${logEntry.timestamp}] ${level}: ${message}`);
    }

    info(message) {
        this.log('INFO', message);
    }

    error(message) {
        this.log('ERROR', message);
    }

    warn(message) {
        this.log('WARN', message);
    }

    getLogs() {
        return this.logs;
    }

    clearLogs() {
        this.logs = [];
    }
}

// ============================================================================
// USAGE EXAMPLES
// ============================================================================

// Eager Singleton
const eager1 = EagerSingleton.getInstance();
const eager2 = EagerSingleton.getInstance();
console.log(eager1 === eager2); // true

// Lazy Singleton
const lazy1 = LazySingleton.getInstance();
const lazy2 = LazySingleton.getInstance();
console.log(lazy1 === lazy2); // true

// Module Singleton
const module1 = ModuleSingleton.getInstance();
const module2 = ModuleSingleton.getInstance();
console.log(module1 === module2); // true

// Modern Singleton
const modern1 = ModernSingleton.getInstance();
const modern2 = ModernSingleton.getInstance();
console.log(modern1 === modern2); // true

// Database Connection
const db1 = DatabaseConnection.getInstance();
const db2 = DatabaseConnection.getInstance();
console.log(db1 === db2); // true

db1.connect('postgresql://localhost:5432/mydb');
db2.query('SELECT * FROM users'); // Works, same instance

// Logger
const logger1 = Logger.getInstance();
const logger2 = Logger.getInstance();
console.log(logger1 === logger2); // true

logger1.info('Application started');
logger2.error('Something went wrong');
logger1.warn('This is a warning');

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        EagerSingleton,
        LazySingleton,
        ModuleSingleton,
        ModernSingleton,
        DatabaseConnection,
        Logger
    };
}

File: intreview_training/design-patterns/creational/singleton/implementations/python/singleton.py
"""
Singleton Pattern - Python Implementations

Multiple approaches to implement Singleton pattern in Python
"""

# ============================================================================
# 1. MODULE-LEVEL SINGLETON (Simplest - Pythonic way)
# ============================================================================
# In Python, modules are singletons by default
# Just create instance at module level
_singleton_instance = None

def get_singleton():
    global _singleton_instance
    if _singleton_instance is None:
        _singleton_instance = SingletonClass()
    return _singleton_instance


# ============================================================================
# 2. DECORATOR APPROACH
# ============================================================================
def singleton(cls):
    """Decorator to make a class a singleton"""
    instances = {}
    
    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    
    return get_instance


@singleton
class DecoratedSingleton:
    def __init__(self):
        self.data = "Decorated Singleton Data"
    
    def get_data(self):
        return self.data


# ============================================================================
# 3. __new__ METHOD APPROACH (Classic Python way)
# ============================================================================
class ClassicSingleton:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(ClassicSingleton, cls).__new__(cls)
            cls._instance.data = "Classic Singleton Data"
        return cls._instance
    
    def get_data(self):
        return self.data


# ============================================================================
# 4. METACLASS APPROACH (Advanced)
# ============================================================================
class SingletonMeta(type):
    """Metaclass for creating singleton classes"""
    _instances = {}
    
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(SingletonMeta, cls).__call__(*args, **kwargs)
        return cls._instances[cls]


class MetaclassSingleton(metaclass=SingletonMeta):
    def __init__(self):
        self.data = "Metaclass Singleton Data"
    
    def get_data(self):
        return self.data


# ============================================================================
# 5. THREAD-SAFE SINGLETON
# ============================================================================
import threading

class ThreadSafeSingleton:
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                # Double-checked locking pattern
                if cls._instance is None:
                    cls._instance = super(ThreadSafeSingleton, cls).__new__(cls)
                    cls._instance.data = "Thread-Safe Singleton Data"
        return cls._instance
    
    def get_data(self):
        return self.data


# ============================================================================
# 6. DATABASE CONNECTION EXAMPLE
# ============================================================================
class DatabaseConnection:
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super(DatabaseConnection, cls).__new__(cls)
                    cls._instance.connected = False
                    cls._instance.connection_string = ""
        return cls._instance
    
    def connect(self, connection_string):
        if self.connected:
            print("Already connected")
            return self
        
        self.connection_string = connection_string
        self.connected = True
        print(f"Connected to: {connection_string}")
        return self
    
    def disconnect(self):
        self.connected = False
        print("Disconnected")
    
    def query(self, sql):
        if not self.connected:
            raise Exception("Not connected to database")
        print(f"Executing: {sql}")
        return {"rows": []}


# ============================================================================
# 7. LOGGER EXAMPLE
# ============================================================================
class Logger:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Logger, cls).__new__(cls)
            cls._instance.logs = []
        return cls._instance
    
    def log(self, level, message):
        log_entry = {
            "timestamp": __import__('datetime').datetime.now().isoformat(),
            "level": level,
            "message": message
        }
        self.logs.append(log_entry)
        print(f"[{log_entry['timestamp']}] {level}: {message}")
    
    def info(self, message):
        self.log("INFO", message)
    
    def error(self, message):
        self.log("ERROR", message)
    
    def warn(self, message):
        self.log("WARN", message)
    
    def get_logs(self):
        return self.logs
    
    def clear_logs(self):
        self.logs = []


# ============================================================================
# USAGE EXAMPLES
# ============================================================================
if __name__ == "__main__":
    # Decorated Singleton
    decorated1 = DecoratedSingleton()
    decorated2 = DecoratedSingleton()
    print(decorated1 is decorated2)  # True
    print(decorated1.get_data())
    
    # Classic Singleton
    classic1 = ClassicSingleton()
    classic2 = ClassicSingleton()
    print(classic1 is classic2)  # True
    print(classic1.get_data())
    
    # Metaclass Singleton
    meta1 = MetaclassSingleton()
    meta2 = MetaclassSingleton()
    print(meta1 is meta2)  # True
    print(meta1.get_data())
    
    # Thread-Safe Singleton
    thread1 = ThreadSafeSingleton()
    thread2 = ThreadSafeSingleton()
    print(thread1 is thread2)  # True
    print(thread1.get_data())
    
    # Database Connection
    db1 = DatabaseConnection()
    db2 = DatabaseConnection()
    print(db1 is db2)  # True
    
    db1.connect("postgresql://localhost:5432/mydb")
    db2.query("SELECT * FROM users")
    
    # Logger
    logger1 = Logger()
    logger2 = Logger()
    print(logger1 is logger2)  # True
    
    logger1.info("Application started")
    logger2.error("Something went wrong")
    logger1.warn("This is a warning")
